### PR TITLE
Matjones/79 hooks refactor

### PR DESCRIPTION
**Pull request by mrjones2014**
Resolves: #79 

@hsheikhali1 sorry this is kinda massive, I randomly decided to add `eslint` in this MR as well 😬

# How to Test

1. Add the following options to `yacltrc.js`:
```js
  preNew: () => console.log("!!!!! PRE NEW"),
  postNew: () => console.log("!!! POST NEW"),
  preValidate: () => console.log("!!! PRE VALIDATE"),
  postValidate: () => console.log("!!! POST VALIDATE"),
  prePrepare: () => console.log("!!! PRE PREPARE"),
  postPrepare: () => console.log("!!! POST PREPARE"),
```
2. Run `yarn build`
3. Run each command (`new`, `validate`, `prepare-release`) from `./dist/index.js` and ensure you see the appropriate `console.log` outputs. Note that since `prepare-release` also runs `validate`, the output for that one should be:
```
!!! PRE PREPARE
!!! PRE VALIDATE
✔ All changelog entries formatted correctly!
!!! POST VALIDATE
✔ CHANGELOG.md updated! Be sure to review the changes before comitting.
!!! POST PREPARE
```
4. Go back to `yacltrc.js` and update the hook functions to `return false`, verify that they now halt command execution
5. Go back to `yacltrc.js` and make all the hook functions `async`, remove the `return false`, then test step 3 again
6. Go back to `yacltrc.js` and keep the hook functions `async` but update them to `return false` again, and verify that they halt command execution

### CODE REVIEW

**Comment by mrjones2014** [comment]
This will definitely be a major release FYI

File: src/cli/commands/prepare-release.ts [hunk 1]
```diff
@@ -42,32 +47,31 @@ export const PrepareReleaseCommand: CommandModule<
...
-      }
-
-      let template: string;
-
-      if (fs.existsSync(argv.changelogTemplate)) {
-        template = fs.readFileSync(argv.changelogTemplate).toString();
-      } else {
-        template = argv.changelogTemplate;
-      }
+      const template = fs.existsSync(argv.changelogTemplate)
```
Comment (mrjones2014) [inline src/cli/commands/prepare-release.ts:67]: Note to self: remove this -- this is now handled by default by the `template-from-file` middleware.

File: src/cli/commands/prepare-release.ts [hunk 2]
```diff
@@ -42,32 +47,31 @@ export const PrepareReleaseCommand: CommandModule<
...
+        : argv.changelogTemplate;
 
-      let releaseNumber: string;
-
-      if (typeof argv.releaseNumber === "string") {
-        releaseNumber = argv.releaseNumber;
-      } else {
-        releaseNumber = argv.releaseNumber();
-      }
+      const releaseNumber =
```
Comment (mrjones2014) [inline src/cli/commands/prepare-release.ts:71]: note to self: this ternary is redundant now, this is handled by the `call-function-args` middleware.

File: src/cli/middleware/call-function-args.ts [hunk 3]
```diff
@@ -1,26 +1,53 @@
...
+import { ActionPrepareReleaseOptions } from "../../actions/prepare-release";
+import { ActionValidateOptions } from "../../actions/validate";
 import { Logger } from "../../utils/logger";
 import { nameof } from "../../utils/nameof";
 import { isFunction } from "../../utils/type-utils";
-import { GlobalArgv } from "../options";
 import { MiddlewareHandler } from "./middleware-handler";
 
-// Record<string, any> allows for any option set whatsoever
+const camelToKebabCase = (str: string): string => {
```
Comment (mrjones2014) [inline src/cli/middleware/call-function-args.ts:10]: move this to a utils file.

**Comment by mrjones2014** [comment]
@hsheikhali1 I tested this pretty thoroughly myself. I'm going to merge so I can get a release out.

**Comment by hsheikhali1** [comment]
Yeah I already tested this too just forgot to update the ticket

**Pull request merged**